analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/ConditionEvaluatesToConstantBase.cs

```diff
@@ -75,16 +75,19 @@ public override ProgramState ConditionEvaluated(SymbolicContext context)
     private bool IsIgnored(ProgramState state, IOperation operation) =>
         operation.Kind is OperationKindEx.Literal
         || IsVarPattern(operation)
-        || IsDiscardPattern(operation)
+        || IsDiscardPattern(state, operation)
         || operation.TrackedSymbol(state) is IFieldSymbol { IsConst: true }
         || operation.Syntax.Ancestors().Any(x => IsInsideUsingDeclaration(x) || IsLockStatement(x));
 
     private static bool IsVarPattern(IOperation operation) =>
         operation.AsIsPattern()?.Pattern.WrappedOperation.AsDeclarationPattern()?.MatchesNull is true;
 
-    private static bool IsDiscardPattern(IOperation operation) =>
-        operation.AsIsPattern() is { } pattern
-        && pattern.Pattern.WrappedOperation.Kind is OperationKindEx.DiscardPattern;
+    private static bool IsDiscardPattern(ProgramState state, IOperation operation) =>
+        operation.Kind is OperationKindEx.DiscardPattern
+        || (operation.AsIsPattern() is { } isPattern
+            && state.ResolveCaptureAndUnwrapConversion(isPattern.Pattern.WrappedOperation) is var subPattern
+            && (subPattern.Kind is OperationKindEx.DiscardPattern
+                || (subPattern.AsRecursivePattern() is {} recursivePattern && recursivePattern.DeconstructionSubpatterns.Any(x => IsDiscardPattern(state, x)))));
```

Comment:
This needs to be `All` not `Any` or it will create a lot of FNs. Add some UTs e.g. like this:
```
name1 = null;
name2 = null;
return (name1, name2) switch
{
(null, null) => 0,  // FN
(null, _) => 1,     // FN
(_, null) => -1,    // FN
(_, _) => Comparer<string>.Default.Compare(name1, name2)
};
```